programs/native/template_signer/src/lib.rs

```diff
@@ -0,0 +1,579 @@
+                ref sig,
+            } => {
+                if msg != pending.msg || !self.template.key.check_sig(&msg, sig) {
+                    return Err(Error::Error);
+                }
+                if self.signed_timeout() {
+                    return Err(Error::Error);
+                }
+                State::ConcurrentSignature {
+                    pending: Some(pending.clone()),
+                    concurrent: msg,
+                    sig: sig.clone(),
+                }
+            }
+            _ => {
+                if self.template_timeout() {
+                    return Err(Error::Error);
+                }
+                if !self.template.key.check_sig(&msg, &sig) {
+                    return Err(Error::Error);
+                }
+                State::ConcurrentSignature {
+                    pending: None,
+                    concurrent: msg,
+                    sig,
+                }
+            }
+        };
+        self.state = nxt;
+        Ok(())
+    }
+    pub fn claim(&mut self, keyed_accounts: &mut Vec<KeyedAccount>) -> Result<(), Error> {
+        //TODO: implement this
+        let nxt = match self.state {
+            State::Requested { .. } => {
+                if !self.requested_timeout() {
+                    return Err(Error::Error);
+                }
+                if self.owner != *keyed_accounts[0].key {
+                    return Err(Error::Error);
+                }
+                let total = keyed_accounts[1].account.tokens;
+                keyed_accounts[0].account.tokens += total;
+                keyed_accounts[1].account.tokens -= total;
+                self.escrow = 0;
+                self.guarantee = 0;
+                //TODO: abstract guarantee and escrow of non token assets
+                State::Claimed
+            }
+            State::Signed { .. } if !self.signed_timeout() => {
+                if self.signer != *keyed_accounts[0].key {
+                    return Err(Error::Error);
+                }
+                let total = keyed_accounts[1].account.tokens;
+                keyed_accounts[0].account.tokens += total;
+                keyed_accounts[1].account.tokens -= total;
+                self.escrow = 0;
+                self.guarantee = 0;
+                //TODO: abstract guarantee and escrow of non token assets
+                State::Claimed
+            }
+            State::ConcurrentSignature { .. } => {
+                if self.owner != *keyed_accounts[0].key {
+                    return Err(Error::Error);
+                }
+                let total = keyed_accounts[1].account.tokens;
+                keyed_accounts[0].account.tokens += total;
+                keyed_accounts[1].account.tokens -= total;
+                self.escrow = 0;
+                self.guarantee = 0;
+                //TODO: abstract guarantee and escrow of non token assets
+                State::Claimed
+            }
+            _ => {
```

Comment:
I think being explicit about all the states is better than a default case.  Makes the program more resilient to future bugs as new states are added